rust/src/libnm_dbus/device.rs

```diff
@@ -481,18 +482,46 @@ fn nm_dev_state_reason_get(
     }
 }
 
+fn nm_dev_is_mac_vtap_get(
+    dbus_conn: &zbus::Connection,
+    obj_path: &str,
+) -> Result<bool, NmError> {
+    let dbus_iface = format!("{}.Macvlan", NM_DBUS_INTERFACE_DEV);
+    let proxy = zbus::Proxy::new(
+        dbus_conn,
+        NM_DBUS_INTERFACE_ROOT,
+        obj_path,
+        &dbus_iface,
+    )?;
+    match proxy.get_property::<bool>("Tab") {
+        Ok(v) => Ok(v),
+        Err(e) => Err(NmError::new(
+            ErrorKind::Bug,
+            format!(
+                "Failed to retrieve Macvlan.Tab(tap) of device {}: {}",
+                obj_path, e
+            ),
+        )),
+    }
+}
+
 pub(crate) fn nm_dev_from_obj_path(
     dbus_conn: &zbus::Connection,
     obj_path: &str,
 ) -> Result<NmDevice, NmError> {
     let (state, state_reason) = nm_dev_state_reason_get(dbus_conn, obj_path)?;
-    Ok(NmDevice {
+    let mut dev = NmDevice {
         name: nm_dev_name_get(dbus_conn, obj_path)?,
         iface_type: nm_dev_iface_type_get(dbus_conn, obj_path)?,
         state,
         state_reason,
         obj_path: obj_path.to_string(),
-    })
+        is_mac_vtap: false,
+    };
+    if dev.iface_type == "macvlan" {
+        dev.is_mac_vtap = nm_dev_is_mac_vtap_get(dbus_conn, obj_path)?;
```

Comment:
hm. This approach with zbus' proxies seems expensive.

No question that blocking calls result in clean, readable code. On the surface at last. It's not very obvious/readable that such functions contain blocking IO. Anyway, it seems expensive and subtly wrong (like all blocking D-Bus calls are fundamentally flawed because they don't preserve the ordering of events on D-Bus -- https://smcv.pseudorandom.co.uk/2008/11/nonblocking/ ).

This function `nm_dev_from_obj_path()` calls three functions, that each create (and initialize) a D-Bus proxy object. That means, the function requires at least 3 blocking D-Bus method calls (that is, round trips). Possibly even more, because initializing a D-Bus proxy object might require more than D-Bus calls (like first an AddMatch or GetNameOwner, dunno)...

The three creations could be avoided in part, by creating the proxy object earlier and passing it to the getter functions. Or even let the caller provide the proxy object already (as the caller might have another use of the proxy object to further). Basically, cache the proxy object and reuse the things.

Reply:
https://docs.rs/zbus/1.9.2/zbus/struct.Proxy.html#current-limitations

ah. OK, so properties are not cached and the name owner is not tracked. I guess that means every `get_property()` call translates to a `Get()` D-Bus call...

Reply:
ok, rethinking...

if the zbus proxy does no caching, then these `get_property()` calls are plain `Get` D-Bus calls (at which point it's not clear what the benefit of the proxy object is, you could just as well make a zbus `call()`). 

Anyway. You need three properties, and you either need 3 `Get()` calls, or one `GetAll()` / `GetManagedObjects()` call. With `GetAll()` (and especially `GetManagedObjects()`) you fetch more than you need, but only have one round-trip. It's not clear which one is better... you anyway will always require at least one roundtrip.

Btw, libnm only makes one `GetManagedObjects()` call at the beginning (to get the initial state). Afterwards, it only relies on the PropertiesChanged signals to get updates. That it needs to maintain a cache of everything, that you might use in the future. The downside is that much of the data you might never actually use and the effort is wasted.

But making 3 `Get()` requests one after the other, is what causes this to be 3 times slower than if you could fire them off in parallel. If you think further and are initializing many `NmDevice`, then each of them requires 3 (or one 1) round trip, and the real gain would be, if you could fire off the initialization of them all in parallel. That would require some sort of `async` implementation...

Reply:
Currently we see no problem on this sync code and we suffered enough in python for the async way. For the getall suggestions, I will use it if someday we noticed a big impact. For the direct call of zbus `call`, again no issue no change.